magma/primitives/lut.py

```diff
@@ -0,0 +1,71 @@
+from typing import Tuple
+
+from hwtypes import BitVector
+
+from magma.bit import Bit
+from magma.bits import Bits
+from magma.bitutils import clog2
+from magma.braid import fork
+from magma.circuit import coreir_port_mapping
+from magma.conversions import from_bits, as_bits
+from magma.interface import IO
+from magma.t import Kind, In, Out
+from magma.generator import Generator2
+
+
+class _CoreIRLUT(Generator2):
+    """
+    Internally used generator for CoreIR LUT primitive
+    """
+
+    def __init__(self, contents: BitVector):
+        n = clog2(len(contents))
+        self.name = f"coreir_lut{str(int(contents))}"
+        self.io = IO(I=In(Bits[n]), O=Out(Bit))
+        self.renamed_ports = coreir_port_mapping
+
+        def _simulate(self, value_store, state_store):
+            in_ = BitVector[n](value_store.get_value(getattr(self, "in")))
+            value_store.set_value(self.out, contents.as_bool_list()[in_])
+        self.simulate = _simulate
+        self.coreir_name = "lutN"
+        self.coreir_lib = "commonlib"
+        self.coreir_genargs = {"N": n}
+        self.coreir_configargs = {"init": contents}
+
+
+def _to_int(value):
+    if isinstance(value, Bit):
+        if not value.const():
+            raise ValueError(f"Unexpected Bit value: {value}")
+        return 1 if value is Bit.VCC else 0
+    elif not isinstance(value, Bits):
+        value = as_bits(value)
+    return int(value)
+
+
+class LUT(Generator2):
+    """
+    Generate a LUT containing entries of a generic Type
+    """
+
+    def __init__(self, T: Kind, n: int, contents: Tuple):
+        """
+        T: The type of each LUT entry (e.g. Bit or Bits[5])
+
+        n: nubmer of entries in the LUT
+
+        contents: the contents of the LUT (a tuple of length `n` containing
+                  elements of type T that are constant)
+        """
+        num_bits_per_entry = T.flat_length()
+        self.io = IO(I=In(Bits[clog2(n)]), O=Out(T))
+
+        contents = tuple(BitVector[num_bits_per_entry](_to_int(c)) for c in
```

Comment:
should we check the type of contents early here? afraid it might lead to some weird/unexpected behavior if wire w/o the check.

Reply:
The type check is done by the BitVector conversion, so if it doesn't satisfy that interface we'll get an error, I think that's a pretty reasonable choice.  We could be more strict (enforcing it's an instance of T), but I would imagine there's case when the user would want to pass a raw BitVector or int rather than constructing the full type.  So the current interface is "can it be converted to a BitVector of the required number of bits", if not we'll get a type error.  This is in the spirit of Python duck typing style (can it act as a BitVector[n]?)

Reply:
That makes sense, and I agree. I think I was originally confused as to why the type annotation for `contents` was a `magma.Tuple` but I realized thats `typing.Tuple` actually :+1: